zkevm-circuits/src/table.rs

```diff
@@ -25,35 +25,65 @@ use keccak256::plain::Keccak;
 use std::array;
 use strum_macros::{EnumCount, EnumIter};
 
-/// Trait used for dynamic tables.  Used to get an automatic implementation of
-/// the LookupTable trait where each `table_expr` is a query to each column at
-/// `Rotation::cur`.
-pub trait DynamicTableColumns {
-    /// Returns the list of advice columns following the table order.
-    fn columns(&self) -> Vec<Column<Advice>>;
-}
-
 /// Trait used to define lookup tables
 pub trait LookupTable<F: Field> {
-    /// Return the list of expressions used to define the lookup table.
-    fn table_exprs(&self, meta: &mut VirtualCells<F>) -> Vec<Expression<F>>;
-}
+    /// Returns the list of ALL the table columns following the table order.
+    fn columns(&self) -> Vec<Column<Any>>;
+
+    /// Returns the list of ALL the table advice columns following the table
+    /// order.
+    fn advice_columns(&self) -> Vec<Column<Advice>> {
+        self.columns()
+            .iter()
+            .map(|&col| col.try_into())
+            .filter_map(|res| res.ok())
+            .collect()
+    }
 
-impl<F: Field, T: DynamicTableColumns> LookupTable<F> for T {
+    /// Retunrs the String annotations associated to each column of the table.
+    fn annotations(&self) -> Vec<String>;
+
+    /// Return the list of expressions used to define the lookup table.
     fn table_exprs(&self, meta: &mut VirtualCells<F>) -> Vec<Expression<F>> {
         self.columns()
             .iter()
-            .map(|column| meta.query_advice(*column, Rotation::cur()))
+            .map(|&column| meta.query_any(column, Rotation::cur()))
             .collect()
     }
+
+    /// Annotates a lookup table by passing annotations for each of it's
+    /// columns.
+    fn annotate_columns(&self, cs: &mut ConstraintSystem<F>) {
+        self.columns()
+            .iter()
+            .zip(self.annotations().iter())
+            .for_each(|(&col, ann)| cs.annotate_lookup_any_column(col, || ann))
+    }
+
+    /// Annotates a lookup table by passing annotations for each of it's
+    /// columns.
+    fn annotate_columns_in_region(&self, region: &mut Region<F>) {
+        self.columns()
+            .iter()
+            .zip(self.annotations().iter())
+            .for_each(|(&col, ann)| region.name_column(|| ann, col))
+    }
 }
 
-impl<F: Field, C: Into<Column<Any>> + Clone, const W: usize> LookupTable<F> for [C; W] {
+impl<F: Field, C: Into<Column<Any>> + Copy, const W: usize> LookupTable<F> for [C; W] {
     fn table_exprs(&self, meta: &mut VirtualCells<F>) -> Vec<Expression<F>> {
         self.iter()
-            .map(|column| meta.query_any(column.clone(), Rotation::cur()))
+            .map(|column| meta.query_any(*column, Rotation::cur()))
             .collect()
     }
+
+    fn columns(&self) -> Vec<Column<Any>> {
+        self.iter().map(|&col| col.into()).collect()
+    }
+
+    fn annotations(&self) -> Vec<String> {
+        vec![]
```

Comment:
The `annotate_columns` and `annotates_columns_in_region` would panic if there are `W` columns, but no annotations, right?

Reply:
I thought about this too, but I think it will not panic.  The reason is that the result of this method is used with `zip` and not `zip_eq`.  And `zip` allows zipping two iterators of different length (so in this case, after the zip it will be an iterator of length 0 so there will be no annotation performed).